### PR TITLE
GB7 inviting users

### PR DESCRIPTION
**Pull request by kris-kolosek**

### CODE REVIEW

File: db/migrate/20180119145644_devise_invitable_add_to_users.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      t.index      :invitations_count
+      t.index      :invitation_token, unique: true # for invitable
+      t.index      :invited_by_id
+    end
+  end
+
+  def down
+    change_table :users do |t|
+      t.remove_references :invited_by, polymorphic: true
+      t.remove :invitations_count, :invitation_limit, :invitation_sent_at, :invitation_accepted_at, :invitation_token, :invitation_created_at
```
Comment (sniffler) [inline db/migrate/20180119145644_devise_invitable_add_to_users.rb]: Line is too long. [141/80]

File: app/models/user.rb [hunk 2]
```diff
@@ -17,6 +17,10 @@ class User < ApplicationRecord
   private
 
   def set_role
-    self.role = Role.new(user: self, role: Role::ADMIN)
+    if User.count.zero?
```
Comment (sniffler) [inline app/models/user.rb]: Use the return of the conditional for variable assignment and comparison.

File: app/controllers/invitations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      format.html do
+        super
+      end
+    end
+  end
+
+  protected
+
+  def update_sanitized_params
+    devise_parameter_sanitizer.permit(:accept_invitation, keys: [:password, :invitation_token])
```
Comment (sniffler) [inline app/controllers/invitations_controller.rb]: Use %i or %I for an array of symbols.<br>Line is too long. [95/80]

File: app/controllers/invitations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  before_action :update_sanitized_params, only: :update
+
+  # PUT /resource/invitation
+  def update
+    respond_to do |format|
+      format.js do
+        invitation_token = Devise.token_generator.digest(resource_class, :invitation_token, update_resource_params[:invitation_token])
+        self.resource = resource_class.where(invitation_token: invitation_token).first
+        resource.skip_password = true
+        resource.update_attributes update_resource_params.except(:invitation_token)
```
Comment (sniffler) [inline app/controllers/invitations_controller.rb]: Line is too long. [83/80]

File: app/controllers/invitations_controller.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+class InvitationsController < Devise::InvitationsController
+  before_action :update_sanitized_params, only: :update
+
+  # PUT /resource/invitation
+  def update
+    respond_to do |format|
+      format.js do
+        invitation_token = Devise.token_generator.digest(resource_class, :invitation_token, update_resource_params[:invitation_token])
+        self.resource = resource_class.where(invitation_token: invitation_token).first
```
Comment (sniffler) [inline app/controllers/invitations_controller.rb]: Line is too long. [86/80]

File: app/controllers/invitations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
+class InvitationsController < Devise::InvitationsController
+  before_action :update_sanitized_params, only: :update
+
+  # PUT /resource/invitation
+  def update
+    respond_to do |format|
+      format.js do
+        invitation_token = Devise.token_generator.digest(resource_class, :invitation_token, update_resource_params[:invitation_token])
```
Comment (sniffler) [inline app/controllers/invitations_controller.rb]: Line is too long. [134/80]

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -4,8 +4,18 @@ class ApplicationController < ActionController::Base
...
   protect_from_forgery with: :exception
   before_action :authenticate_user!
+  before_action :configure_permitted_parameters, if: :devise_controller?
 
   def current_profile
     @_profile ||= current_user.profile
   end
+
+  protected
+
```
Comment (sniffler) [inline app/controllers/application_controller.rb]: Trailing whitespace detected.

File: app/models/user.rb [hunk 8]
```diff
@@ -17,6 +17,10 @@ class User < ApplicationRecord
...
+      self.role = Role.new(user: self, role: Role::ADMIN)
```
Comment (salac-kolosek) [inline app/models/user.rb]: Make role like this build_role(status: Role::Admin).

File: app/views/devise/invitations/_new.html.erb [hunk 9]
```diff
@@ -0,0 +1,14 @@
+<div class="invitation-container">
+  <button type="button" class="close" data-dismiss="modal" aria-label="Close">
+    <span aria-hidden="true">&times;</span>
+  </button>
+  <h2 class="title">Invite a New User</h2>
+  <%= form_for resource, :as => resource_name, :url => invitation_path(resource_name), :remote => true, :html => {:method => :post} do |f| %>
```
Comment (salac-kolosek) [inline app/views/devise/invitations/_new.html.erb]: I think you don't need all of this params (:as, :url,: method) for form_for helper. Plus use Rails 5 sintax with remote: true for example.

File: app/views/users/index.html.erb [hunk 10]
```diff
@@ -1,21 +1,41 @@
...
-            <%= button_to "Delete", {:controller => "registrations", :action => 'destroy'}, data: { confirm: "Are you sure you want to delete #{user.profile.full_name}?" }, method: 'delete', class: 'btn delete' %>
+        <% if !user.invitation_token.present? %>
+          <% if !user.profile.present? %>
+            <div class='member'> 
+              <b class='name'><%= user.email %></b>
+            </div>
           <% end %>
-        </div>
+        <% end %>
+        <% if user.profile.present? %>
```
Comment (salac-kolosek) [inline app/views/users/index.html.erb]: Defensive programming detected :)

File: app/views/users/index.html.erb [hunk 11]
```diff
@@ -1,21 +1,41 @@
...
   <div class='label'>Acitve users</div>
   <div class='members-list'>
     <% @users.each do |user| %>
       <%= link_to profile_path(user.profile) do %>
-        <div class='member'> 
-           <%= image_tag edit_image(user.profile.avatar_url), size:'50x50', class: 'avatar-icon' %>
-          <b class='name'><%= user.profile.full_name %></b>
-          <% if current_user.role.admin? %>
-            <%= button_to "Delete", {:controller => "registrations", :action => 'destroy'}, data: { confirm: "Are you sure you want to delete #{user.profile.full_name}?" }, method: 'delete', class: 'btn delete' %>
+        <% if !user.invitation_token.present? %>
```
Comment (salac-kolosek) [inline app/views/users/index.html.erb]: Defensive programming detected :) At least make this one if statement.

File: app/views/users/index.html.erb [hunk 12]
```diff
@@ -1,21 +1,41 @@
...
+            </div>
           <% end %>
-        </div>
+        <% end %>
+        <% if user.profile.present? %>
+          <div class='member'> 
+             <%= image_tag edit_image(user.profile.avatar_url), size:'50x50', class: 'avatar-icon' %>
+            <b class='name'><%= user.profile.full_name %></b>
+            <% if current_user.role.admin? %>
+              <%= button_to "Delete", {:controller => "registrations", :action => 'destroy'}, data: { confirm: "Are you sure you want to delete #{user.profile.full_name}?" }, method: 'delete', class: 'btn delete' %>
```
Comment (salac-kolosek) [inline app/views/users/index.html.erb]: Use rails 5 syntax.

File: app/models/user.rb (hunk 8 above)
Comment (kris-kolosek) [reply app/models/user.rb]: I will change this, thank you

**Pull request closed without merging**